logrotate.c

```diff
@@ -409,29 +409,51 @@ static int setSecCtx(int fdSrc, const char *src, char **pPrevCtx)
     return 0;
 }
 
-static int setSecCtxByName(const char *src, const struct logInfo *log, char **pPrevCtx)
+static int setSecCtxByName(const char *src, char **pPrevCtx)
 {
-    int hasErrors = 0;
 #ifdef WITH_SELINUX
-    int fd;
+    char *srcCtx;
+    *pPrevCtx = NULL;
 
     if (!selinux_enabled)
         /* pretend success */
         return 0;
 
-    fd = open_logfile(src, log, 0);
-    if (fd < 0) {
-        message(MESS_ERROR, "error opening %s: %s\n", src, strerror(errno));
-        return 1;
+    /* read security context of src */
+    if (lgetfilecon_raw(src, &srcCtx) < 0) {
+        if (errno == ENOTSUP)
+            /* pretend success */
+            return 0;
+
+        message(MESS_ERROR, "getting file context %s: %s\n", src,
+                strerror(errno));
+        return selinux_enforce;
+    }
+
+    /* save default security context for restoreSecCtx() */
+    if (getfscreatecon_raw(pPrevCtx) < 0) {
+        message(MESS_ERROR, "getting default context: %s\n", strerror(errno));
+        freecon(srcCtx);
+        return selinux_enforce;
     }
-    hasErrors = setSecCtx(fd, src, pPrevCtx);
-    close(fd);
+
+    /* set default security context to match src */
+    if (setfscreatecon_raw(srcCtx) < 0) {
+        message(MESS_ERROR, "setting default context to %s: %s\n", srcCtx,
+                strerror(errno));
+        freecon(*pPrevCtx);
+        *pPrevCtx = NULL;
+        freecon(srcCtx);
+        return selinux_enforce;
+    }
+
+    message(MESS_DEBUG, "set default create context to %s\n", srcCtx);
+    freecon(srcCtx);
```

Comment:
This code is copy/pasted from `setSecCtx()`.  It would be better to move it to a new function that would be called by both `setSecCtx()` and `setSecCtxByName()` so that it does not need to be maintained at two places.

Reply:
good idea, updated.